content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,84 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).
+- **Tests should check not only valid inputs but also invalid inputs and edge cases** (if the problem allows them). `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
+
+
+## Random Tests
+
+Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
+
+- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.
+- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
+- In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.
+- **The order of random tests should be non-deterministic** especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria). Not only should inputs be randomized, but their order as well. It should not be the case that, for example, 10 valid, random inputs are tested in the first step followed by 10 invalid ones. The order of tested scenarios should be unpredictable too, and one possible technique is to generate and collect a set of random inputs for all required scenarios and shuffle them.
```

Comment:
this part needs a small rework:

"the order of the random tests" here is actually about "the order of the __outputs__". Hence saying clearly that "the kinds of outputs shouldn't be grouped together when random testing" seems far more... clear.

Reply:
Right, I just had no good idea how to word it. I keep this in mind.

Reply:
this way?

```suggestion
- **Random tests shouldn't be grouped by kind of outputs or behaviours**, especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria), or when it comes to error checking (solution throwing an exception in some scpecific situations). Hence, the order of tested scenarios should be unpredictable too. One possible way to achieve this is to generate and collect a set of random inputs for all required scenarios and shuffle them before the actual tests.
```

note: `scenarii`?

Reply:
> note: `scenarii`?

Nah, [Wikitionary](https://en.wiktionary.org/wiki/scenarii#English) says that such usage is rare in modern English.